### PR TITLE
Send correct http return code for avatar-endpoint

### LINKED ISSUE #28725: Avatar endpoints return incorrect status code
### Description

The API endpoints `/user/avatar` for creating/deleting an avatar has a single status code: 204. This is incorrect and doesn't work well with automation like ansible.

If the avatar is created successfully, then the endpoint should return 201 (Created). If no change was made then it should return something else. If there was an error it should return something else.

Same for the delete endpoint - it currently returns one status code.

Also note that the current behaviour (204) is different to the other endpoints.

### Gitea Version

1.21.3

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

docker

### Database

SQLite

### PR DESCRIPTION
**Pull request by zokkis**
Send correct http status code for */avatar api-endpoints (Fix #28725)

### CODE REVIEW

File: routers/api/v1/user/avatar.go [hunk 1]
```diff
@@ -26,8 +26,10 @@ func UpdateAvatar(ctx *context.APIContext) {
 	//   schema:
 	//     "$ref": "#/definitions/UpdateUserAvatarOption"
 	// responses:
-	//   "204":
-	//     "$ref": "#/responses/empty"
```
Comment (delvh) [inline routers/api/v1/user/avatar.go:30]: This line should always be kept.

**Review by delvh — COMMENTED** [review]
Hmm… I'm not particularly sure if this is a good idea.
After all, we promise API stability, and this is by definition a breaking change.
However, if the current return value is the wrong one as per the standard and there isn't much reliance on the return being exactly 204, then I'm slightly in favor of this change.

**Comment by zokkis** [comment]
> Hmm… I'm not particularly sure if this is a good idea. After all, we promise API stability, and this is by definition a breaking change. However, if the current return value is the wrong one as per the standard and there isn't much reliance on the return being exactly 204, then I'm slightly in favor of this change.

As written on [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses) for updating or creating it should send 201

But because the body is empty, so the 204 is not 100% wrong

**Comment by KN4CK3R** [comment]
Tests need adjustment too.

**Comment by lafriks** [comment]
> As written on [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses) for updating or creating it should send 201
> 

This is not true, 201 should only be returned when a resource has been created not updated. For updates it should return 200 or 204 depending whether returning something in response or not

**Comment by zokkis** [comment]
> > As written on [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses) for updating or creating it should send 201
> 
> This is not true, 201 should only be returned when a resource has been created not updated. For updates it should return 200 or 204 depending whether returning something in response or not

Should be better now

File: routers/api/v1/org/avatar.go [hunk 2]
```diff
@@ -43,12 +47,25 @@ func UpdateAvatar(ctx *context.APIContext) {
...
+	if hasAvatar {
+		ctx.Status(http.StatusOK)
+	} else {
+		ctx.Status(http.StatusCreated)
+	}
+	_, err = ctx.Write(avatarData)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "Write", err)
+		return
+	}
```
Comment (KN4CK3R) [inline routers/api/v1/org/avatar.go:68]: Why does the upload now respond with the avatar?

Comment (zokkis) [reply routers/api/v1/org/avatar.go:68]: https://github.com/go-gitea/gitea/pull/29185#issuecomment-1950910348

Mdn specification is written `the newly created items are returned in the body of the message` for 201

Comment (KN4CK3R) [reply routers/api/v1/org/avatar.go:68]: Where?
> [201 Created](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201)
> The request succeeded, and a new resource was created as a result. This is typically the response sent after POST requests, or some PUT requests.

I would not expect a server to responsd with my 1gb file I just uploaded. I understand it for endpoints like creating a user to return the user json but not for file uploads.

Comment (zokkis) [reply routers/api/v1/org/avatar.go:68]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201

But the User is not the Object which is created

**Comment by zokkis** [comment]
any updates? Or should I close it?

File: routers/api/v1/org/avatar.go [hunk 3]
```diff
@@ -72,6 +89,7 @@ func DeleteAvatar(ctx *context.APIContext) {
 	err := user_service.DeleteAvatar(ctx, ctx.Org.Organization.AsUser())
 	if err != nil {
 		ctx.Error(http.StatusInternalServerError, "DeleteAvatar", err)
+		return
```
Comment (lunny) [inline routers/api/v1/org/avatar.go:92]: This is useful.

File: routers/api/v1/org/avatar.go (hunk 2 above)
Comment (lunny) [reply routers/api/v1/org/avatar.go:68]: Maybe don't return any values like before?

File: routers/api/v1/repo/avatar.go [hunk 4]
```diff
@@ -48,12 +52,24 @@ func UpdateAvatar(ctx *context.APIContext) {
...
+		return
 	}
 
-	ctx.Status(http.StatusNoContent)
+	if hasAvatar {
+		ctx.Status(http.StatusOK)
+	} else {
+		ctx.Status(http.StatusCreated)
+	}
+	_, err = ctx.Write(avatarData)
```
Comment (lunny) [inline routers/api/v1/repo/avatar.go:68]: As above.

File: routers/api/v1/repo/avatar.go [hunk 5]
```diff
@@ -82,6 +98,7 @@ func DeleteAvatar(ctx *context.APIContext) {
 	err := repo_service.DeleteAvatar(ctx, ctx.Repo.Repository)
 	if err != nil {
 		ctx.Error(http.StatusInternalServerError, "DeleteAvatar", err)
+		return
```
Comment (lunny) [inline routers/api/v1/repo/avatar.go:101]: This is useful.

**Comment by lunny** [comment]
> any updates? Or should I close it?

I don't think return the avatar to client is necessary. And some bug fixes are very important which should be fixed ASAP.

**Comment by lafriks** [comment]
I agree, I don't see added benefits for returning uploaded avatar back to client

**Pull request closed without merging**